kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoEditorWorkflow.kt

```diff
@@ -36,50 +38,84 @@ data class TodoRow(
   val done: Boolean = false
 )
 
-data class TodoRendering(
+sealed class TodoAction : WorkflowAction<Nothing, TodoEditorOutput> {
+  object GoBackClicked : TodoAction()
+
+  sealed class ListAction : TodoAction() {
+    abstract val list: TodoList
+
+    class TitleChanged(
+      override val list: TodoList,
+      val newTitle: String
+    ) : ListAction()
+
+    class DoneClicked(
+      override val list: TodoList,
+      val index: Int
+    ) : ListAction()
+
+    class TextChanged(
+      override val list: TodoList,
+      val index: Int,
+      val newText: String
+    ) : ListAction()
+
+    class DeleteClicked(
+      override val list: TodoList,
+      val index: Int
+    ) : ListAction()
+  }
+
+  override fun Mutator<Nothing>.apply(): TodoEditorOutput {
+    return when (this@TodoAction) {
+      is GoBackClicked -> Done
+      is TitleChanged -> ListUpdated(list.copy(title = newTitle))
+      is DoneClicked -> ListUpdated(list.updateRow(index) { copy(done = !done) })
+      is TextChanged -> ListUpdated(list.updateRow(index) { copy(text = newText) })
+      is DeleteClicked -> ListUpdated(list.removeRow(index))
+    }
+  }
+}
+
+class TodoRendering(
   val list: TodoList,
-  val onEvent: (TodoEvent) -> Unit
+  val onTitleChanged: (title: String) -> Unit,
+  val onDoneClicked: (index: Int) -> Unit,
+  val onTextChanged: (index: Int, text: String) -> Unit,
+  val onDeleteClicked: (index: Int) -> Unit,
+  val onGoBackClicked: () -> Unit
 )
 
 sealed class TodoEditorOutput {
   data class ListUpdated(val newList: TodoList) : TodoEditorOutput()
   object Done : TodoEditorOutput()
 }
 
-sealed class TodoEvent {
-  data class TitleChanged(val title: String) : TodoEvent()
-  data class DoneClicked(val index: Int) : TodoEvent()
-  data class TextChanged(
-    val index: Int,
-    val text: String
-  ) : TodoEvent()
-
-  data class DeleteClicked(val index: Int) : TodoEvent()
-  object GoBackClicked : TodoEvent()
-}
-
 class TodoEditorWorkflow : StatelessWorkflow<TodoList, TodoEditorOutput, TodoRendering>() {
 
   override fun render(
     input: TodoList,
     context: RenderContext<Nothing, TodoEditorOutput>
   ): TodoRendering {
+    val rawSink = context.makeSink<TodoAction>()
+
+    // Make event handling idempotent.
+    var eventFired = false
+    val sink = object : Sink<TodoAction> {
+      override fun send(value: TodoAction) {
+        if (eventFired) return
+        eventFired = true
+        rawSink.send(value)
+      }
+    }
```

Comment:
File an issue with a TODO here to implement the same "if the sink type is the same across render passes, let multiple calls to the same sink instance work" behavior that Swift has.

Reply:
https://github.com/square/workflow/issues/541